### PR TITLE
[MPDX-8360] Fix changing contact filters reloading the page

### PR DESCRIPTION
**Pull request by canac**
## Description

The root cause of buggy routing behavior throughout the app is that changes to the router don't seem to take effect immediately. For example, when clearing filters, one `useEffect` removes `filters` from the query and another `useEffect` removes `searchTerm` from the query. However, when the second effect runs, the change to the filters haven't propagated yet, and it ends up adding the filters back to the URL.

My solution, which has seemed to work well on the contacts page, is to put state that affects the URL (like filters and the search query) into a `useState`. Changes to that state will update immediately. The context wrapper then updates the URL as the state changes. Because the state variables are the source of truth instead of the URL, delays in the URL updating don't cause synchronization issues anymore.

I would like to eventually extract the logic in the wrapper component into another url filter context that can be used in contacts, tasks, partner giving analysis, and responsibility centers. But I'd like to wait on that for now and verify that this approach works better.

There is still buggy behavior on the appeals page when switching between list and flow mode. I don't believe that it is any more buggy than it was before, and it can be addressed in a future PR. It does seem like I've fixed the bugginess on the appeals list view when clicking the "Given", "Received", etc, filters ([MPDX-8239](https://jira.cru.org/browse/MPDX-8239)).

Selecting contacts and then clicking map view to only view those contacts doesn't work, but it didn't work before, so that can be done in a future PR.

I have also done some cleanup like combining variables and remove the extra parameters from `setContactFocus`. At one point, I thought that debouncing hook was buggy and wrote more tests to verify its implementation. The hook was actually correct, but I'm leaving the tests since that hook didn't have test coverage before.

[MPDX-8360](https://jira.cru.org/browse/MPDX-8360)

## Checklist:

- [x] I have given my PR a title with the format "MPDX-(JIRA#) (summary sentence max 80 chars)"
- [x] I have applied the appropriate labels. (_Add the label "On Staging" to get the branch automatically merged into staging._)
- [x] I have requested a review from another person on the project

### CODE REVIEW

**Review by dr-bizz — COMMENTED** [review]
As we spoke about there seems to be a bug when switching from `list` to `flows` etc.

**Review by dr-bizz — CHANGES_REQUESTED** [review]
This is looking great! The PR is pretty confusing as there are a lot of large changes which took me a while to review them. I will probably have to have another pass at it since I probably missed some things.

The code is looking good, I did spot an error on the Tasks filters when filtering, which will need to be fixed before we push this into production.

https://github.com/user-attachments/assets/7803c6fe-bb84-440a-9a01-8845368d5f98

**Comment by canac** [comment]
> The code is looking good, I did spot an error on the Tasks filters when filtering, which will need to be fixed before we push this into production.

@dr-bizz Thanks for catching that. It should be fixed now.

Sorry that this code is still really hairy and hard to follow because the contacts context is split into context and wrapper components. Also, non-contacts pages use the contacts context. It could use some refactoring to pull out URL filters logic and contact panel logic into reusable context. Then we would only need contacts context on the contacts pages. But that will have to be a future improvement.

**Review by dr-bizz — APPROVED** [review]
I've tested this and it looks great

File: pages/accountLists/[accountListId]/tools/appeals/AppealsWrapper.tsx [hunk 1]
```diff
@@ -81,24 +84,31 @@ export const AppealsWrapper: React.FC<Props> = ({ children }) => {
     });
   };
 
+  const setActiveFilters: AppealsContextProps['setActiveFilters'] = (value) => {
+    const filters = typeof value === 'function' ? value(activeFilters) : value;
```
[the remark this answers is not in the document]
Comment (canac) [reply pages/accountLists/[accountListId]/tools/appeals/AppealsWrapper.tsx:88]: It's just because `Dispatch<SetStateAction<_>>` means that you can pass a value or a function to the setter. But I don't think we're actually passing in a function. I could have also changed the type of `setActiveFilters` to `(filters: ContactFilterSetInput & TaskFilterSetInput) => void`. I might actually do that.

File: pages/accountLists/[accountListId]/contacts/ContactsWrapper.tsx [hunk 2]
```diff
@@ -95,7 +98,7 @@ export const ContactsWrapper: React.FC<Props> = ({ children }) => {
       newQuery.searchTerm = encodeURIComponent(searchTerm);
     }
     return newQuery;
-  }, [accountListId, contactId, viewMode, activeFilters, searchTerm]);
+  }, [contactId, viewMode, activeFilters, searchTerm]);
```
Comment (dr-bizz) [inline pages/accountLists/[accountListId]/contacts/ContactsWrapper.tsx:101]: Should `addViewMode` be added as a dependency?

Comment (dr-bizz) [reply pages/accountLists/[accountListId]/contacts/ContactsWrapper.tsx:101]: I guess it won't change much

**Review by dr-bizz — APPROVED** [review]
The code is looking good, I'm going to review it on preview

**Comment by dr-bizz** [comment]
⚠️ Might be best to push this live on Monday when we are all around for the next few days.

**Comment by dr-bizz** [comment]
My bad; clicked the wrong button.

**Comment by dr-bizz** [comment]
...I've caused quite a bit of drama 😅

**Comment by canac** [comment]
After waiting for the branch to finish deleting, I recreated it again, so the old preview link will work again after it finishes building.
